### PR TITLE
DTensor + collectives

### PR DESCRIPTION
**Pull request by d4l3k**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123685
* #123684

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### CODE REVIEW

File: torch/distributed/_tensor/ops/collectives_ops.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+import torch
+from torch.distributed._tensor.op_schema import OpSchema, OutputSharding
+from torch.distributed._tensor.ops.utils import register_prop_rule
+from torch.distributed._tensor.placement_types import DTensorSpec
+
+_c10d_functional = torch.ops._c10d_functional
+
+
+@register_prop_rule(_c10d_functional.all_to_all_single.default)
```
Comment (yifuwang) [inline torch/distributed/_tensor/ops/collectives_ops.py:12]: Do we already have similar rules for other collective?

Comment (wanchaol) [inline torch/distributed/_tensor/ops/collectives_ops.py:12]: Can you elaborate why you need this?

DTensor works on top of collectives, so it calls collectives under the hood to transform a tensor sharding layout. Letting DTensor to work with the collective op directly is a pretty confusing behavior IMO. For example: why we keep the original sharding if DTensor runs with the alltoall_single op?

Comment (d4l3k) [reply torch/distributed/_tensor/ops/collectives_ops.py:12]: Nope, there's no other rules

Comment (d4l3k) [reply torch/distributed/_tensor/ops/collectives_ops.py:12]: The idea here is to allow mixing parallelism strategies -- i.e. supporting using DTensor for tensor parallelism while being able to use the collectives for higher level operations.

I was thinking about how to combine tensor parallelism with ring attention -- with this we end up rotating the kv blocks but still preserve the sharding information for tensor parallelism

I'm not sure this is the best direction to go by any means. I did this as an experiment after thinking about how you could compose DTensor and collectives after chatting with @wconstab last week

**Pull request closed without merging**